compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1140,7 +1140,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         // are the same function and their parameters have a LUB.
                         match self.commit_if_ok(|_| {
                             self.at(cause, self.param_env).lub(
-                                DefineOpaqueTypes::No,
+                                DefineOpaqueTypes::Yes,
```

Comment:
This needs a test for it. I think we could create a test exercising this by coercing closures to fn ptrs when they have no captures.